rmw_zenoh_cpp/src/detail/rmw_publisher_data.cpp

```diff
@@ -0,0 +1,454 @@
+  options.attachment = z_bytes_map_as_attachment(&map);
+
+  if (shmbuf.has_value()) {
+    zc_shmbuf_set_length(&shmbuf.value(), data_length);
+    zc_owned_payload_t payload = zc_shmbuf_into_payload(z_move(shmbuf.value()));
+    ret = zc_publisher_put_owned(z_loan(pub_), z_move(payload), &options);
+  } else {
+    // Returns 0 if success.
+    ret = z_publisher_put(
+      z_loan(pub_),
+      reinterpret_cast<const uint8_t *>(msg_bytes),
+      data_length,
+      &options);
+  }
+  if (ret) {
+    RMW_SET_ERROR_MSG("unable to publish message");
+    return RMW_RET_ERROR;
+  }
+
+  return RMW_RET_OK;
+}
+
+///=============================================================================
+rmw_ret_t PublisherData::publish_serialized_message(
+  const rmw_serialized_message_t * serialized_message,
+  std::optional<zc_owned_shm_manager_t> & /*shm_manager*/)
+{
+  eprosima::fastcdr::FastBuffer buffer(
+    reinterpret_cast<char *>(serialized_message->buffer), serialized_message->buffer_length);
+  rmw_zenoh_cpp::Cdr ser(buffer);
+  if (!ser.get_cdr().jump(serialized_message->buffer_length)) {
+    RMW_SET_ERROR_MSG("cannot correctly set serialized buffer");
+    return RMW_RET_ERROR;
+  }
+
+  std::lock_guard<std::mutex> lock(mutex_);
+
+  z_owned_bytes_map_t map =
+    rmw_zenoh_cpp::create_map_and_set_sequence_num(sequence_number_++, gid_);
+
+  if (!z_check(map)) {
+    // create_map_and_set_sequence_num already set the error
+    return RMW_RET_ERROR;
+  }
+  auto free_attachment_map = rcpputils::make_scope_exit(
+    [&map]() {
+      z_bytes_map_drop(z_move(map));
+    });
+
+  const size_t data_length = ser.get_serialized_data_length();
+
+  // The encoding is simply forwarded and is useful when key expressions in the
+  // session use different encoding formats. In our case, all key expressions
+  // will be encoded with CDR so it does not really matter.
+  z_publisher_put_options_t options = z_publisher_put_options_default();
+  options.attachment = z_bytes_map_as_attachment(&map);
+
+  // Returns 0 if success.
+  int8_t ret = z_publisher_put(
+    z_loan(pub_),
+    serialized_message->buffer,
+    data_length,
+    &options);
+
+  if (ret) {
+    RMW_SET_ERROR_MSG("unable to publish message");
+    return RMW_RET_ERROR;
+  }
+
+  return RMW_RET_OK;
+}
+
+///=============================================================================
+std::size_t PublisherData::guid() const
+{
+  std::lock_guard<std::mutex> lock(mutex_);
+  return entity_->guid();
+}
+
+///=============================================================================
+liveliness::TopicInfo PublisherData::topic_info() const
+{
+  std::lock_guard<std::mutex> lock(mutex_);
+  return entity_->topic_info().value();
+}
+
+///=============================================================================
+const uint8_t * PublisherData::gid() const
+{
+  std::lock_guard<std::mutex> lock(mutex_);
+  return gid_;
```

Comment:
Because this is returning a raw pointer, this is probably not thread-safe.

That said, from what I can tell there is only one caller of this API, from `rmw_get_gid_for_publisher`.  And there, what we are doing is copying the data into an out parameter.

That suggests to me that this function should be something like:

```
const void PublisherData::copy_gid(rmw_gid_t * gid) const
{
  std::lock_guard<std::mutex> lock(mutex_);
  memcpy(gid->data, gid_, RMW_GID_STORAGE_SIZE);
}
```

(and then update the caller down below as needed).  That way we are doing the copy under the lock, which should be safe.

Reply:
Good idea! Also had to do something similar for copying the GID into the attachment map https://github.com/ros2/rmw_zenoh/pull/278/commits/4edbee2a038869d71529243748f0534a2258bae5